rfcs/20211207-softmax-v2/README.md

```diff
@@ -0,0 +1,144 @@
+Mentioned primitives are still executed in floating-point data types such as `f32`, `bf16` or `f16`, but it usually happens that followed by primitive would use quantized output from mentioned primitives.
+It means that every time such situation happens, a reorder from floating-point type to integer type appears.
+
+In Transformer-like models, including BERT, softmax layer appears in every Encoder and Decoder block.
+The number of blocks may vary but we may consider a base case when 12 of each are present resulting in 24 reorders bringing their overhead.
+
+Reorder can easily be incorporated into softmax operation, or any other operation, by letting it store the answer in requested data type.
+The problem that `data_desc` does not allow to do that and the only possible solution is to extend softmax descriptor to accept additional memory descriptor.
+Quantized output of mentioned primitives would not be meaningful in terms of operation essence, but this is what's used in production for inference and was tested accordingly from accuracy point of view.
+
+### Motivation
+
+Major motivation is free performance improvements for Transformer-like models for int8 inference.
+Additional motivation is to eliminate `logsoftmax` primitive API which was a result of preserving ABI for users back in the days.
+
+### Additional Feature Requests
+
+oneDNN team was also asked to provide several features through softmax primitives.
+One of them was temperature argument, which is supported only in MXNet.
+
+It was mentioned in Python API though, e.g., R API (first Google link for MXNet doc) does not have it.
+None of other frameworks are supporting this parameter. Following documentation references were used:
+[Tensorflow](https://www.tensorflow.org/api_docs/python/tf/nn/softmax),
+[PyTorch](https://pytorch.org/docs/stable/generated/torch.nn.Softmax.html),
+[OpenVINO](https://docs.openvino.ai/latest/openvino_docs_ops_activation_SoftMax_1.html).
+In its core, it would divide the data by `temperature`, see [MXNet documentation](https://mxnet.apache.org/versions/master/api/python/docs/api/npx/generated/mxnet.npx.softmax.html) for details.
+Same can be achieved if precede layer would apply `binary_div` post-op with `mask` of `0`.
+Due to the lack of broader adoption and existence of simple work around, such feature will not be added to oneDNN softmax primitive.
+
+Another long-lasting request is so-called `masked` softmax.
+It comes from Transformer-based models when Decoder applies Encoder-Decoder Attention layer.
+At this point the machine is not allowed to see future tokens of the sentence, that's why this "mask" should hide the future starting from the processed token.
+Mask makes softmax input data shorter.
+
+Due to the nice property of softmax that data on the left end of range does not contribute much or at all (depending on the range) to final answer and likely vanishes, there is a trick to perform this optimization on user side.
```

Comment:
What does the user do to perform the trick? I don't understand, but that may be my technical ignorance.

Reply:
Yeah, makes sense to me, added suggested note, thanks.